### PR TITLE
Update protobuf toolchain & CICD

### PR DESCRIPTION
**Pull request by hochiw**
Ref oursky/likecoin-chain#40 oursky/likecoin-chain#222 oursky/likecoin-chain#213

- Adopt ignite-cli (starport) for codegen from proto, drops gen_proto.sh
- Adopt `buf mod` to manage proto deps
- Adopt `buf format` to enforce proto code formatting
- Enable protobuf linting on CI

Currently linting is disabled and not fixed for iscn protos to avoid breaking changes, we should fix and re-enable lints when we make changes to iscn module in the future

### CODE REVIEW

**Comment by chihimng** [comment]
why? version controlling dependency sources sound bad to me; why can't we read from golang's local cache?

even if we really cannot read from cache in gopath, do a `go mod vendor` and git ignore the vendor folder instead

manually checking in sdk source every time we update dependencies is a hurdle and prone to error...

**Comment by chihimng** [comment]
we can skip linting sdk protos if it works; if it's just that the linter cannot reference third party files, please find a way to fix it without checking those into version control 🙏

**Comment by hochiw** [comment]
took reference from
https://github.com/terra-money/core/tree/main/third_party/proto
https://github.com/tharsis/evmos/tree/main/third_party/proto

but let's go with the vendor approach

**Comment by hochiw** [comment]
> we can skip linting sdk protos if it works; if it's just that the linter cannot reference third party files, please find a way to fix it without checking those into version control 🙏

in this case maybe skip `proto-check-breaking` entirely as it checks against an upstream branch in which it never works if the third party files are not in the branch

**Comment by hochiw** [comment]
@chihimng updated :adore2x:

File: proto/buf.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+version: v1
+deps:
+  - buf.build/cosmos/cosmos-sdk
```
Comment (chihimng) [inline proto/buf.yaml:3]: 👍

File: proto/buf.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+lint:
+  use:
+    - DEFAULT
+  except:
```
Comment (chihimng) [inline proto/buf.yaml:7]: do we have a reference for excluding these?

& can we follow the sdk one? https://github.com/cosmos/cosmos-sdk/blob/master/proto/buf.yaml

we can exclude iscn if it fails, but better enforce standards to likenft

**Review by elise-ng — CHANGES_REQUESTED** [review]

File: proto/buf.yaml (hunk 2 above)
Comment (hochiw) [reply proto/buf.yaml:7]: let's exclude iscn first

**Review by elise-ng — APPROVED** [review]

**Comment by chihimng** [comment]
handle this after v2.0 milestone

**Comment by chihimng** [comment]
Ref #229, we better add the extra ignores & fix proto folder structure for iscn modules (should be `/proto/likechain/iscn` for `likechain.iscn` package) when submit upstream

cc @hochiw

**Review by elise-ng — CHANGES_REQUESTED** [review]
> Ref https://github.com/oursky/likecoin-chain/pull/229, we better add the extra ignores & fix proto folder structure for iscn modules (should be /proto/likechain/iscn for likechain.iscn package) when submit upstream

needs follow up

**Review by elise-ng — APPROVED** [review]

**Comment by chihimng** [comment]
Let's merge #236 into feat/nft first, then cherry pick changes to here for upstream?

**Comment by chihimng** [comment]
cherry picked commits, and updated pr description for upstream submission

**Comment by chihimng** [comment]
@hochiw please see if my changes are ok?

**Comment by chihimng** [comment]
side note, seems docker is the easier way to use buf for now, it somehow installs another copy of go via brew (i use asdf), and for linux there is no package manager support at all: https://docs.buf.build/installation

**Comment by chihimng** [comment]
We might want to revert the folder structure change and mute the lint error instead, it's not working well with starport

**Comment by rickmak** [comment]
Oh, Close it first?

**Comment by rickmak** [comment]
Sent at https://github.com/likecoin/likecoin-chain/pull/78

**Pull request closed without merging**